flow/raster_cache.cc

```diff
@@ -26,18 +26,30 @@ RasterCacheResult::RasterCacheResult(sk_sp<SkImage> image,
                                      const char* type)
     : image_(std::move(image)), logical_rect_(logical_rect), flow_(type) {}
 
-void RasterCacheResult::draw(SkCanvas& canvas, const SkPaint* paint) const {
+void RasterCacheResult::draw(SkCanvas& canvas,
+                             const SkPaint* paint,
+                             const SkMatrix* matrix) const {
   TRACE_EVENT0("flutter", "RasterCacheResult::draw");
   SkAutoCanvasRestore auto_restore(&canvas, true);
+  auto device_total_matrix = canvas.getTotalMatrix();
   SkIRect bounds =
-      RasterCache::GetDeviceBounds(logical_rect_, canvas.getTotalMatrix());
+      RasterCache::GetDeviceBounds(logical_rect_, device_total_matrix);
   FML_DCHECK(
       std::abs(bounds.size().width() - image_->dimensions().width()) <= 1 &&
       std::abs(bounds.size().height() - image_->dimensions().height()) <= 1);
   canvas.resetMatrix();
   flow_.Step();
-  canvas.drawImage(image_, bounds.fLeft, bounds.fTop, SkSamplingOptions(),
-                   paint);
+  if (matrix == nullptr) {
+    canvas.drawImage(image_, bounds.fLeft, bounds.fTop, SkSamplingOptions(),
+                     paint);
+  } else {
+    FML_DCHECK(matrix->isScaleTranslate());
+    SkRect dest_rect;
+    matrix->mapRect(&dest_rect, logical_rect_);
+    SkRect dest_rect_2;
+    device_total_matrix.mapRect(&dest_rect_2, dest_rect);
+    canvas.drawImageRect(image_, dest_rect_2, SkSamplingOptions(), paint);
```

Comment:
You'd want to pump through the sampling options from the original matrix filter here I think.

Reply:
Yes.

Reply:
Also, don't try to do this variant of the cache blit with bounding rectangles. Supply the rendering call with the transform you need and let the graphics engine do the work. There's some transform math to be done here (the original matrix produced this cache image and now there's a new matrix to render under), but you could allow for any kind of transform matrix here.

Basically I think we can assert that the entry was cached with the same matrix we are rendering with or we wouldn't have gotten this far. The only difference are the translation components. Then we want to apply the given matrix parameter on top of that, and translate it to the current canvas's translation component and draw.

Reply:
The ideal would be for there to be a drawImageQuad() where you specify the 4 points of the quad into which to render the image, but there is drawAtlas that lets you specify a scaled parallelogram into which to render the image. It would work for any scaled and rotated transform. Beyond that there is drawVertices which allows you to specify 2 triangles into which to map the bitmap.